metrics: seperater metrics as internal and external for slo-controller and koordlet

zwzhang0107 opened the pull request:
### Ⅰ. Describe what this PR does

<!--
- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)
-->
seperate koordlet/slo-controller metrics as internal and external
ExternalHTTPPath: for end users, such as batch resource util (also compatible with old URL /metircs for koordlet）
InternalHTTPPath: for cluster admin, such as performance of components.
DefaultHTTPPath: merge all internal, external and controller runtime metrics.

### Ⅱ. Does this pull request fix one issue?

<!--If so, add "fixes #xxxx" below in the next line, for example, fixes #15. Otherwise, add "NONE" -->

### Ⅲ. Describe how to verify it

### Ⅳ. Special notes for reviews

### V. Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `make test`

saintube commented on cmd/koord-manager/main.go:
```diff
@@ -186,3 +195,14 @@ func setRestConfig(c *rest.Config) {
 	}
 }
+
+func installHTTPHandler(mgr ctrl.Manager) error {
```
does this need to move into the extensions/options directory for internal modifications?

zwzhang0107 replied:
not necessary since metrics.InternalRegistry/ExternalRegistry are global

saintube reviewed (commented):
Need to update the deploy templates under the `/config` since the metrics paths are going to change.

ZiMengSheng commented:
/milestone v1.5

zwzhang0107 commented:
> Need to update the deploy templates under the `/config` since the metrics paths are going to change.

the url path has not defined yet both in config and charts

eahydra commented:
Is it really necessary to split it into two metrics urls? Users also need to configure two prometheus configurations for slo-controller, and this indicator must be distinguished when adding indicators in future code. Moreover, only two components, koord-manager and koordlet, are involved here. Do scheduler/descheduler also need to be followed? I don't think such a distinction should be made, and it is not a common or recommended practice.

zwzhang0107 commented:
> Is it really necessary to split it into two metrics urls? Users also need to configure two prometheus configurations for slo-controller, and this indicator must be distinguished when adding indicators in future code. Moreover, only two components, koord-manager and koordlet, are involved here. Do scheduler/descheduler also need to be followed? I don't think such a distinction should be made, and it is not a common or recommended practice.

@eahydra  Yes, it is necessary. any extended metric will be charged(ref https://help.aliyun.com/zh/ack/ack-managed-and-ack-dedicated/user-guide/dynamic-resource-overcommitment?spm=a2c4g.11186623.0.0.74d118ab3vDxaM#ef961016a2uds), and for end users they only need care about `external` metrics, this will cost a lot if we do not separate internal and external after we add new metrics about performance.

scheduler/descheduler should also split if they have the same senario in future.

eahydra commented:
> > Is it really necessary to split it into two metrics urls? Users also need to configure two prometheus configurations for slo-controller, and this indicator must be distinguished when adding indicators in future code. Moreover, only two components, koord-manager and koordlet, are involved here. Do scheduler/descheduler also need to be followed? I don't think such a distinction should be made, and it is not a common or recommended practice.
> 
> @eahydra Yes, it is necessary. any extended metric will be charged(ref https://help.aliyun.com/zh/ack/ack-managed-and-ack-dedicated/user-guide/dynamic-resource-overcommitment?spm=a2c4g.11186623.0.0.74d118ab3vDxaM#ef961016a2uds), and for end users they only need care about `external` metrics, this will cost a lot if we do not separate internal and external after we add new metrics about performance.
> 
> scheduler/descheduler should also split if they have the same senario in future.

This seems to be solved through the reable_config mechanism of prometheus. Only need to configure the metrics regex and add action=keep.

zwzhang0107 commented:
> reable_config

this is not supported for 

> > > Is it really necessary to split it into two metrics urls? Users also need to configure two prometheus configurations for slo-controller, and this indicator must be distinguished when adding indicators in future code. Moreover, only two components, koord-manager and koordlet, are involved here. Do scheduler/descheduler also need to be followed? I don't think such a distinction should be made, and it is not a common or recommended practice.
> > 
> > 
> > @eahydra Yes, it is necessary. any extended metric will be charged(ref https://help.aliyun.com/zh/ack/ack-managed-and-ack-dedicated/user-guide/dynamic-resource-overcommitment?spm=a2c4g.11186623.0.0.74d118ab3vDxaM#ef961016a2uds), and for end users they only need care about `external` metrics, this will cost a lot if we do not separate internal and external after we add new metrics about performance.
> > scheduler/descheduler should also split if they have the same senario in future.
> 
> This seems to be solved through the reable_config mechanism of prometheus. Only need to configure the metrics regex and add action=keep.

prometheus.yaml is not configurable for end user by default in product like ACK

saintube commented:
> ### Ⅰ. Describe what this PR does
> seperate koordlet/slo-controller metrics as internal and external /external-metrics: for end users, such as batch resource util /metrics: for cluster admin, such as performance of components.
> 
> ### Ⅱ. Does this pull request fix one issue?
> ### Ⅲ. Describe how to verify it
> ### Ⅳ. Special notes for reviews
> ### V. Checklist
> * [ ]  I have written necessary docs and comments
> * [ ]  I have added necessary unit tests and integration tests
> * [ ]  All checks passed in `make test`

It seems the external metrics are using the path `/metrics`.

zwzhang0107 commented:
> > ### Ⅰ. Describe what this PR does
> > seperate koordlet/slo-controller metrics as internal and external /external-metrics: for end users, such as batch resource util /metrics: for cluster admin, such as performance of components.
> > ### Ⅱ. Does this pull request fix one issue?
> > ### Ⅲ. Describe how to verify it
> > ### Ⅳ. Special notes for reviews
> > ### V. Checklist
> > 
> > * [ ]   I have written necessary docs and comments
> > * [ ]   I have added necessary unit tests and integration tests
> > * [ ]   All checks passed in `make test`
> 
> It seems the external metrics are using the path `/metrics`.

updated

saintube reviewed (approved):
/lgtm

FillZpp reviewed (approved):
/lgtm

zwzhang0107 commented:
/approve
